Managers/MapManager.gd

```diff
@@ -51,27 +51,162 @@ func create_map(map_floors_width: Array[int] = map_width_array) -> MapBase:
+			var _parent_y: int = current_room.room_position.y - 1
+			if _parent_y >= 0 && _parent_x >= 0 && _parent_x < map.rooms[_parent_y].size() && map.rooms[_parent_y][_parent_x] != null:
+				_parent_events.append(map.rooms[_parent_y][_parent_x].room_event.get_event_name())
+
+		# Check if current room has the same type as one of its parent room
+		if current_room_event_name in _parent_events:
+			return false
+
+
+	## Rule 3: There must be at least 2 room types among destinations of Rooms that have 2 or more Paths going out. [br]
+	## Since the events of the rooms on the right of current room ((x+i,y), i=1,2,...) have not been generated yet, [br]
+	## we don't need to take them into account. We only check neighbors on the left: (x-2,y), (x-1,y) and (x,y) [br]
+	## Note: We make sure the two leftmost and rightmost rooms of the floor have different events, [br]
+	## since they could be the only destinations of a room from the previous floor (happens when current floor's size [br]
+	## is smaller or equal than previous)
+	var _neighbors_events: Array[String] = []
+	var delta_xs: Array[int] = []
+
+	# If rightmost room of floor
+	if current_room.room_position.x == map.rooms[current_room.room_position.y].size() - 1 or map.rooms[current_room.room_position.y][current_room.room_position.x+1] == null:
+		delta_xs = [-1, 0]
+	else:
+		delta_xs = [-2, -1, 0]
+		
+	# Get the neighbor and current rooms types
+	for delta_x: int in delta_xs:
+		var _neighbor_x: int = current_room.room_position.x + delta_x
+		var _neighbor_y: int = current_room.room_position.y
+		if _neighbor_x >= 0 && _neighbor_x < map.rooms[_neighbor_y].size() && map.rooms[_neighbor_y][_neighbor_x] != null:
+			_neighbors_events.append(map.rooms[_neighbor_y][_neighbor_x].room_event.get_event_name())
+
+	# If our current room has no neighbors on the left, ignore this rule
+	if _neighbors_events.size() >= 2:
+		var _unique: Array[String] = []
+
+		# Remove duplicates
+		for event in _neighbors_events:
+			if not _unique.has(event):
+				_unique.append(event)
+		
+		# If there is only one event type among all rooms, redraw event
+		if _unique.size() == 1:
+			return false
+
+
+	## Rule 4: No Heal rooms two floors before Boss
+	if (current_room_event_name == "heal" && current_room.room_position.y == map.rooms.size() - 3):
+		return false
+
+	return true
+
+
+## Draws room type randomly, with each type having a set probability [br]
+## Algorithm taken from this: https://stackoverflow.com/a/1761646
+func draw_room_type() -> EventBase:
+	var sum: int = 0
+	for proba in GlobalVar.EVENTS_PROBABILITIES:
+		sum += proba
+	var _rand_type_number: int = randi_range(0, sum-1)
+
+	for _rand_type_index: int in range(GlobalVar.EVENTS_CLASSIFICATION.size()):
+		if _rand_type_number < GlobalVar.EVENTS_PROBABILITIES[_rand_type_index]:
+			return GlobalVar.EVENTS_CLASSIFICATION[_rand_type_index].new()
+
+		_rand_type_number -= GlobalVar.EVENTS_PROBABILITIES[_rand_type_index]
+	
+	# This shouldn't be reached
```

Comment:
use the `push_error` to show an error in the console, and return a monster event by default
better have an event we didn't want rather than no event at all in a reachable room